Blog post for CLI AI additions (#14970)
pulumi/docs: content/blog/cli-ai-extensions/index.md

---
title: "Pulumi Updates, Explained: AI-Powered Features in Pulumi CLI"
date: 2025-05-15T10:00:00-05:00
draft: false
meta_desc: AI is helping cloud engineers understand deployment issues 
meta_image: meta.png
math: true
authors:
    - artur-laksberg
    - mikhail-shilkov
    - simon-howe
tags:
    - Copilot
    - ai
    - iac
    - pulumi update

---

> ## Introducing AI-Powered insights in Pulumi CLI

Reviewer: IMO this intro goes a bit too far in trying to build a common arc for these two features. I would keep it to a single paragraph, and then move the other two to corresponding sections.

Additionally, our new AI-powered update explanation feature serves as a valuable sanity check before deployment. It presents a plain-language summary of exactly what changes will occur, helping you verify that the planned modifications align with your intentions. This visibility is particularly valuable when managing complex infrastructure, as it helps identify unexpected changes or potential drift between your code and deployed resources. The system also proactively flags potential issues such as security vulnerabilities or destructive updates that might impact your production environment.

## Diagnostics summarization

Consider the following scenario where a program fails at runtime. While the standard diagnostics provide precise and detailed information, they can be time-consuming to analyze. Our AI-powered summary identifies the problem and its location in your source code, reducing time to resolution and helping you address issues efficiently:

<div style="text-align: center; width: 100%; margin: 0 auto;">
    <img src="summarize2.png" alt="" style="width: 100%;">

> </div>

Reviewer 2: Yeah, trying to load up an example that is better than a TS error.

<div style="text-align: center; width: 100%; margin: 0 auto;">
    <img src="summarize2.png" alt="" style="width: 100%;">
    <figcaption>
        <i>Auto-summarization of pulumi preview</i>
    </figcaption>
</div>

## Update explanation

We've introduced a new "explain" menu item in the CLI. When selected, this tool analyzes and provides a clear summary of all changes in your pending infrastructure update:

<div style="text-align: left; width: 80%; margin: 0; padding: 0;">
    <img src="menu.png" alt="" style="width: 100%;">
</div>

The image below shows an example of the tool's output. The explanation provides a concise summary of the pending changes and highlights any potential concerns related to your update, giving you greater visibility before deployment:

<div style="text-align: center; width: 100%; margin: 0 auto;">
    <img src="explain.png" alt="" style="width: 100%;">
    <figcaption>
        <i>Auto-summarization of pulumi preview</i>
    </figcaption>
</div>

## What's next

> The feature is currently in preview. You will need to add the --copilot flag to your pulumi CLI command to enable it:

Author: Maybe mention `--copilot` earlier in the post, or else most people will not see it? 🤔

---
title: "Pulumi Updates, Explained: AI-Powered Features in Pulumi CLI"
date: 2025-05-21
draft: false
meta_desc: AI-powered CLI features simplify deployments by translating complex changes into clear explanations and providing actionable guidance for resolving errors.
meta_image: meta.png
authors:
    - artur-laksberg
    - mikhail-shilkov
    - simon-howe
tags:
    - copilot
    - ai
    - iac
    - cli
social:
    twitter: "🤖 Introducing AI-powered features in Pulumi CLI! Get plain-language explanations of infrastructure changes and actionable insights for error resolution. Try it today with --copilot flag! #Pulumi #AI #IaC"
    linkedin: "Excited to announce AI-powered features in the Pulumi CLI! ✨ Our new 'explain' feature translates complex infrastructure changes into clear, human-readable summaries, while our AI error analysis transforms cryptic cloud provider error messages into actionable guidance. Enable with --copilot flag and let us know what you think!"
---

> Pulumi CLI now features AI capabilities that translate complex infrastructure changes into clear, human-readable explanations and help resolve deployment errors with actionable guidance. Enable these preview features with the `--copilot` flag.

Reviewer 3: Should we make it more clear this is "powered by Pulumi Copilot"?

Also, should we link to http://pulumi.com/copilot for folks who want to learn more? Maybe a bullet at the ending CTAs, "Learn more about Pulumi Copilot"?

This clear explanation identifies root causes and provides actionable guidance, transforming dense technical output into concise explanations. It reduces time to resolution by pinpointing both the problem and the exact command needed to fix it, eliminating the need to search through documentation or guess.

## What's next

The AI-powered features are currently in preview. You can enable them by adding the `--copilot` flag to your Pulumi CLI commands:

```bash
pulumi preview --copilot
pulumi up --copilot
```

During this preview period, we're focused on:

- Refining explanations based on user feedback
- Adding new capabilities like suggesting fixes for common errors
- Improving performance to ensure minimal impact on command execution time

We will continue to monitor usage patterns and incorporate your feedback before enabling these features by default.

## Share your feedback

We're excited to see how these AI-powered insights enhance your infrastructure deployment workflow! Please share your experiences with us:

> - Connect with us in the [Pulumi Community Slack](https://slack.pulumi.com)

Reviewer 3: On this note, how many copilot commands do I get as a free user? What about Team? Do we need to explain that, because it's powered by Pulumi Copilot, this is limited to the Pulumi Cloud backend?

---

> title: "Pulumi Updates, Explained: AI-Powered Features in Pulumi CLI"

Reviewer 3: Maybe just me, but this parses a little funny. Any reason not to just say

```suggestion
title: "AI-Powered Infrastructure as Code Features in Your CLI"
```

Author: The intent was to say that pulumi updates are now explained with AI, but maybe it came out a little awkward. 
Maybe simply "AI-Powered Features in Pulumi CLI"?  ("Infrastructure as Code" is applicable to everything we do and feels a bit redundant here)
